interfaces/sourcegen/sourcegen/_types.py

```diff
@@ -0,0 +1,72 @@
+# This file is part of Cantera. See License.txt in the top-level directory or
+# at https://cantera.org/license.txt for license and copyright information.
+
+from abc import ABCMeta, abstractmethod
+from dataclasses import dataclass, fields
+import os
+
+
+def _unpack(obj):
+    return (getattr(obj, field.name) for field in fields(obj))
+
+
+@dataclass
+class Param:
+    """ Represents a function parameter """
+
+    p_type: str
+    name: str
+
+
+    @staticmethod
+    def parse(c_param: str):
+        parts = c_param.strip().rsplit(' ', 1)
+        if len(parts) == 2:
+            return Param(*parts)
+
+
+    def __iter__(self):
+        return _unpack(self)
+
+
+@dataclass
+class Func:
+    """ Represents a function parsed from a C header file """
+
+    ret_type: str
+    name: str
+    params: list[Param]
+
+
+    @staticmethod
+    def parse(c_func: str):
+        lparen = c_func.index('(')
+        rparen = c_func.index(')')
+        front = c_func[0:lparen].split()
+
+        params = (Param.parse(p) for p in c_func[lparen+1:rparen].split(','))
+        params = [p for p in params if p]
+
+        ret_type = front[-2]
+        name = front[-1]
+        return Func(ret_type, name, params)
+
+
+    def __iter__(self):
+        return _unpack(self)
+
+
+class SourceGeneratorBase(metaclass=ABCMeta):
+    """ Specifies the interface of a language-specific SourceGenerator  """
+
+    @abstractmethod
+    def __init__(self, out_dir: str, config: dict):
+        pass
```

Comment:
As setting `self.out_dir` and `self.config` is required by subclasses, it may as well be implemented here in a concrete method that subclasses can also call if they add their own initializer.

Reply:
I did go with this, but then reversed for a couple of reasons:
- The C# source generator now uses a typed Config class, rather than a loosey-goosey dict.
- The base class doesn't use these attr's, but putting the setter there _does_ obscure it from the derived class
  - Putting it in the `__init__` of the child class makes it explicit where the attrs came from.

Reply:
> * The base class doesn't use these attr's, but putting the setter there _does_ obscure it from the derived class

I don't see the problem with this? The super-class `__init__` is called automatically if the subclass doesn't override it, and if the subclass does override it, then the subclass should call `super().__init__(self, ...)` in its own `__init__`. That's a really common pattern in Python.

I just think that since this is _required_ for the class to function at all, not including it in the superclass will just create boilerplate that will get repeated in every subclass that doesn't need to be there, particularly if there's no other reason to define a subclass `__init__` at all. Or, it's likely to be forgotten if it's not available from the superclass.

Reply:
As of now, `SourceGenerator` (the as-renamed base class) is more like a pure interface; it has no implemented behavior whatsoever. That’s why I don’t think it should set any attributes in the init method. And you’ll see that the `CSharpSourceGenerator` shouldn’t simply pass the raw dict because it’s using the "typed config" class. Plus the `__init__` sets merely two attribute, hardly a high burden of boiler plate to toss around. (But this admittedly coming from Java experience where dozens of lines of boilerplate are common in many scenarios). And of course I think it fits the Zen of Python axioms of “Explicit is better than implicit” and “Simple is better than complex”.

If in the future, we did julia and were like, wow look `CSharpSourceGenerator` and `JuliaSourceGenerator` do a lot of the same work; then yes let’s extract that behavior to the base class and toss things in via the init. But I honestly don’t forsee that happening because the actual work of scaffolding code is indeed so specific to the target language. So, I do think we should leave `SourceGenerator` purely virtual.

Reply:
> If in the future, we did julia and were like, wow look `CSharpSourceGenerator` and `JuliaSourceGenerator` do a lot of the same work; then yes let’s extract that behavior to the base class and toss things in via the init. But I honestly don’t forsee that happening because the actual work of scaffolding code is indeed so specific to the target language.

I agree with this logic, both for this particular instance and for other questions about the `sourcegen` interface. I'm almost certain that trying to implement a new language interface using it will reveal both some common patterns that can be extracted out of the C# generator and probably some cases that require additional flexibility from the existing common code. But I think that's completely fine given that this is an internal tool, and not a library that needs to get locked into a particular stable API.

Reply:
I'm not convinced. These two parameters are required and it seems like they're going to be required for every interface. It will be super easy to forget them, so I still think they should be set in the base class. 

Another way to say this is, if the base class is entirely virtual, why bother having it at all? If we want flexibility, then we should just get rid of the inheritance altogether.

Reply:
There will likely be a  need/desire to modify this API as we add more and more interfaces. @speth, I am also interested in making sure this fits your desires for Cantera/enhancements#39